docs/cross-chain-payments-and-execution.md

```diff
@@ -0,0 +1,77 @@
+}
+
+```
+
+Whenever a fully signed UserOp containing a `PaymentChainInfo` is submitted to chain, it forces a challenge on that chain. This means that once you have a fully signed UserOp you know you can always use it to get your funds via challenge. In the happy path, we would expect participants to just exchange signed states afterwards, so they can discard the UserOp.
+
+## Cross-chain Payment Example
+
+Let's say we have Alice, Bob,and Irene. Alice and Irene have a BridgeWallet on chain A, Bob and Irene have a BridgeWallet on Chain B. Alice and Irene have both signed a state with a balance of `[A:5,I:5]`, and bob and Irene have likewise have a signed state with a balance of`[B:5,I:5]`. Let's say Alice wants to pay Bob.
+
+1. Alice creates two new unsigned states `[A:4,I:6]` and `[B:6,I:4]`
+2. Alice includes them in a UserOp(wrapped in a `PaymentChainInfo`), and signs the UserOp. She sends this to Bob and Irene.
+3. Bob and Irene validate the UserOp and also sign it, and broadcast it.
+4. At this point the UserOp is fully signed and can be used to force a challenge on either chain, by submitting the signed UserOp to the entrypoint on either chain.
+
+# Cross-chain Execution
+
+We can also extend this idea to cross chain execution. Instead of embedding a payment state in the `UserOp` we embed information to make an on-chain function call. If the `UserOp` is submitted to the `entrypoint` the function will be executed **if the chainId matches the chain**.
+
+We embed this information using a `ExecuteChainInfo`
+
+```sol
+/// Contains information to execute a function call
+struct ExecuteChainInfo {
+  uint chainId;
+  address entrypoint;
+  address dest;
+  uint value;
+  bytes callData;
+  address owner;
+  address intermediary;
+}
+```
+
+A UserOp can contain multiple `ExecuteChainInfo`s, meaning you can have "atomic" cross chain execution (once the UserOp is fully signed, you're guaranteed you can trigger the execution on either chain). A UserOp can also contain `PaymentChainInfo`s and `ExecuteChainInfos` allowing to pay for cross-chain execution.
+
+# Multihop
+
+Since a UserOp can contain multiple `PaymentChainInfo`s and `ExecuteChainInfos`, this approach can be used for multi-hop execution or payments. We just require the `owner` and `intermediary` of every BridgeWallet involved to sign the UserOp.
+
+# Replay Attacks
+
+It's important that we're not vulnerable to replay attacks, where a `UserOp` is submitted again using a different entrypoint or chain.
+
+If we're working on one chain, it's fairly easy to prevent replay attacks. The `userOpHash` [provided by the Entrypoint](https://github.com/magmo/Bridge-Wallet/blob/ad6d24fa2435f449751d1b61e24d12faff1f83a9/contracts/core/EntryPoint.sol#L298) to `validateUserOp` is hashed against the current chain and entrypoint. This means that if the UserOp is run against a different chain or entrypoint, there will be a different `userOpHash` causing all the signature checks to fail.
```

Comment:
With [this change](https://github.com/magmo/Bridge-Wallet/pull/1/commits/c4829f5763967725576751055c721a6dcf98e949) I don't think it's too sub-optimal. The `validateCrossChain` would be generating one hash, and then checking 2 signatures for every `BridgeWallet`.

IE: If we have a multi-hop UserOp with Alice,Irene,Iris,Bob our signatures would like this : `[AliceSig,IreneSig,IreneSig,IrisSig,IrisSig,BobSig]`.   This could also be optimized so we avoid submitting duplicate signatures, meaning we'd have a signature per participant.